File: backbone/src/test/java/org/researchstack/backbone/ui/permissions/PermissionResultTest.java
package org.researchstack.backbone.ui.permissions;

import android.Manifest;


import org.junit.Test;
import org.robolectric.util.Pair;

import java.util.ArrayList;
import java.util.List;

import static android.content.pm.PackageManager.PERMISSION_DENIED;
import static android.content.pm.PackageManager.PERMISSION_GRANTED;
import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.MatcherAssert.assertThat;

public class PermissionResultTest {

    private PermissionResult subject;

    @Test
    public void testContainsPermission() {
        String permission = Manifest.permission.CAMERA;

        subject = permissionsWith()
                .add(permission, true)
                .build();


        assertThat(subject.containsResult("not a real permission"), is(false));
        assertThat(subject.containsResult(permission), is(true));
    }

    @Test
    public void testIsGranted_isTrue() {
        String permission = Manifest.permission.CAMERA;

        subject = permissionsWith()
                .add(permission, true)
                .build();


        assertThat(subject.isGranted(permission), is(true));
    }

    @Test
    public void testIsGranted_isFalse() {
        String permission = Manifest.permission.CAMERA;

        subject = permissionsWith()
                .add(permission, false)
                .build();


        assertThat(subject.isGranted(permission), is(false));
    }

    private ResultBuilder permissionsWith() {
        return new ResultBuilder();
    }

    class ResultBuilder {
        List<Pair<String, Integer>> pairs = new ArrayList<>();

        public ResultBuilder add(String permission, boolean isGranted) {
            pairs.add(Pair.create(permission, isGranted? PERMISSION_GRANTED : PERMISSION_DENIED));
            return this;
        }

        public PermissionResult build() {
            String [] permissions = new String[pairs.size()];
            int [] grantResults = new int[pairs.size()];

            for(int i=0; i < permissions.length; i++) {
                permissions[i] = pairs.get(i).first;
                grantResults[i] = pairs.get(i).second;
            }

            return new PermissionResult(permissions, grantResults);
        }
    }
}

File: backbone/src/test/java/org/researchstack/backbone/ui/permissions/PermissionResultTest.kt
package org.researchstack.backbone.ui.permissions

import android.Manifest
import android.content.pm.PackageManager
import org.hamcrest.CoreMatchers
import org.hamcrest.MatcherAssert
import org.junit.Test
import java.util.ArrayList

class PermissionResultTest {
    private var subject: PermissionResult? = null

    @Test
    fun testContainsPermission() {
        val permission = Manifest.permission.CAMERA
        subject = permissionsWith()
            .add(permission, true)
            .build()
        MatcherAssert.assertThat(subject!!.containsResult("not a real permission"), CoreMatchers.`is`(false))
        MatcherAssert.assertThat(subject!!.containsResult(permission), CoreMatchers.`is`(true))
    }

    @Test
    fun testIsGranted_isTrue() {
        val permission = Manifest.permission.CAMERA
        subject = permissionsWith()
            .add(permission, true)
            .build()
        MatcherAssert.assertThat(subject!!.isGranted(permission), CoreMatchers.`is`(true))
    }

    @Test
    fun testIsGranted_isFalse() {
        val permission = Manifest.permission.CAMERA
        subject = permissionsWith()
            .add(permission, false)
            .build()
        MatcherAssert.assertThat(subject!!.isGranted(permission), CoreMatchers.`is`(false))
    }

    private fun permissionsWith(): ResultBuilder {
        return ResultBuilder()
    }

    internal inner class ResultBuilder {
        var pairs: MutableList<Pair<String, Int>> = ArrayList()

        fun add(permission: String, isGranted: Boolean): ResultBuilder {
            pairs.add(Pair(permission, if (isGranted) PackageManager.PERMISSION_GRANTED else PackageManager.PERMISSION_DENIED))
            return this
        }

        fun build(): PermissionResult {
            val permissions = arrayOfNulls<String>(pairs.size)
            val grantResults = IntArray(pairs.size)

            for (i in permissions.indices) {
                permissions[i] = pairs[i].first
                grantResults[i] = pairs[i].second
            }

            return PermissionResult(permissions, grantResults)
        }
    }
}